### PR TITLE
[BEAM-588] Add MemoryReporter to python-sdk

### PR DESCRIPTION
**Pull request by yk5**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by yk5** [comment]
R:@aaltay 

Can you review? Thanks!

File: sdks/python/apache_beam/utils/profiler.py [hunk 1]
```diff
@@ -67,3 +67,77 @@ def __exit__(self, *args):
...
+    return self
+
+  def __exit__(self, *args):
+    self.stop()
+
+  def start(self):
+    if self._enabled:
+      return
+    if not self._hpy:
+      logging.warning('guppy is not installed; MemoryReporter not available.')
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: I think warning once in `__init__` is enough. This could be simplified to
```
if self._enabled or not self._hpy:
  return
```
Similarly for the `report_once`

File: sdks/python/apache_beam/utils/profiler.py [hunk 2]
```diff
@@ -67,3 +67,77 @@ def __exit__(self, *args):
...
+        if not self._enabled:
+          return
+        self.report_once()
+        self._timer = Timer(self._interval_sec, report_with_interval)
+        self._timer.start()
+
+      self._timer = Timer(self._interval_sec, report_with_interval)
+
+    except ImportError:
+      logging.warning('guppy is not installed; MemoryReporter not available.')
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Use `warnings.warn`. It by default prevents duplicate warnings.

File: sdks/python/apache_beam/utils/profiler.py [hunk 3]
```diff
@@ -67,3 +67,77 @@ def __exit__(self, *args):
...
+      <do something>
+      # this will report continuously with 30 seconds between reports.
+    mr.stop()
+
+  Or simply:
+    with MemoryReporter(interval_sec=100):
+      <do some thing>
+
+  Also it could report on demand without continuous reporting.
+    mr = MemoryReporter()  # default interval 60s but not activated.
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Is it `started` or `activated`?

File: sdks/python/apache_beam/utils/profiler.py [hunk 4]
```diff
@@ -67,3 +67,77 @@ def __exit__(self, *args):
...
+    <do something>
+    mr.report_once()
+  """
+
+  def __init__(self, interval_sec=60.0):
+    # guppy might not have installed. Set up the reporter only when guppy is
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Could you add a comment line about, `guppy` not being available on all platforms.

File: sdks/python/apache_beam/utils/profiler.py [hunk 5]
```diff
@@ -67,3 +67,77 @@ def __exit__(self, *args):
...
+    if not self._hpy:
+      logging.warning('guppy is not installed; MemoryReporter not available.')
+      return
+    self._enabled = True
+    self._timer.start()
+
+  def stop(self):
+    if not self._enabled:
+      return
+    self._timer.cancel()
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Unless you recreate `self._timer`, next call to `start` will fail.

Maybe `start` should create the `Timer` object instead of `__init__`

File: sdks/python/apache_beam/utils/profiler.py [hunk 6]
```diff
@@ -67,3 +67,77 @@ def __exit__(self, *args):
...
+    self._enabled = False
+
+  def report_once(self):
+    if not self._hpy:
+      logging.warning('guppy is not installed; MemoryReporter not available.')
+      return
+    report_start_time = time.time()
+    heap_profile = self._hpy().heap()
+    logging.info('*** MemoryReport Heap:\n %s', heap_profile)
+    logging.info('*** MemoryReport took %.1f sec.',
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Do we need the second log statement? Could it be removed or maybe become a debug log?

Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: I find it useful, when MemoryReporter takes like 20 seconds for pretty big profiles. Merged into one line, but if you think it's better to remove, I could remove.

File: sdks/python/apache_beam/utils/profiler.py (hunk 5 above)
Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: Done

File: sdks/python/apache_beam/utils/profiler.py (hunk 1 above)
Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: Done

File: sdks/python/apache_beam/utils/profiler.py (hunk 2 above)
Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: done.

File: sdks/python/apache_beam/utils/profiler.py (hunk 4 above)
Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: Done.

File: sdks/python/apache_beam/utils/profiler.py (hunk 3 above)
Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: Done.

**Comment by yk5** [comment]
Thanks for review! Can you PTAL? @aaltay

File: sdks/python/apache_beam/utils/profiler.py [hunk 7]
```diff
@@ -134,10 +139,8 @@ def stop(self):
...
   def report_once(self):
     if not self._hpy:
-      logging.warning('guppy is not installed; MemoryReporter not available.')
       return
     report_start_time = time.time()
     heap_profile = self._hpy().heap()
-    logging.info('*** MemoryReport Heap:\n %s', heap_profile)
-    logging.info('*** MemoryReport took %.1f sec.',
-                 time.time() - report_start_time)
+    logging.info('*** MemoryReport Heap:\n %s\n MemoryReport took %.1f sec',
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Trying this out, Output looks like this

`...`
`<47 more rows. Type e.g. '_.more' to view.>`

Should we actually iterate over this to print the full picture? Otherwise it prints only the first 10 items.

Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Could you spell out `sec`s to `second`s?

Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: It goes more than thousand for complicated pipelines. I would dump to disk instead, as cpu profiler does. But I'd like to do it in a separate PR, when it's plugged into DirectRunner using pipeline options. WDYT?

Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: Done

**Comment by yk5** [comment]
Thanks for review!

I would plug into DirectRunner using PipelineOptions in a separate PR. Would it be okay?
@aaltay

File: sdks/python/apache_beam/utils/profiler.py (hunk 6 above)
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Let's keep it. If you find it useful, others will do too.

File: sdks/python/apache_beam/utils/profiler.py (hunk 7 above)
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Sounds good. Please add a note to JIRA issue about this so that we do not forget about it.

**Comment by aaltay** [comment]
LGTM. Thank you @yk5

File: sdks/python/apache_beam/utils/profiler.py (hunk 6 above)
Comment (yk5) [inline sdks/python/apache_beam/utils/profiler.py]: Yes, I'm keeping the info; just that it doesn't need to be in a separate logging line, so that I merged into the previous logging.

**Comment by yk5** [comment]
Thanks! I think it's ready to merge. :)
